src/Data/Store/Impl.hs

```diff
@@ -165,27 +165,42 @@ decodeIO = decodeIOWith peek
 -- exceptions, and taking a 'Peek' to run. It is an exception to not
 -- consume all input.
 decodeIOWith :: Peek a -> BS.ByteString -> IO a
-decodeIOWith mypeek bs = do
-    (offset, x) <- decodeIOPortionWith mypeek bs
-    let remaining = BS.length bs - offset
-    if remaining > 0
-        then throwIO $ PeekException remaining "Didn't consume all input."
-        else return x
+decodeIOWith mypeek (BS.PS x s len) =
+    withForeignPtr x $ \ptr0 ->
+        let ptr = ptr0 `plusPtr` s
+        in decodeIOWithFromPtr mypeek ptr len
 {-# INLINE decodeIOWith #-}
 
 -- | Similar to 'decodeExPortionWith', but runs in the 'IO' monad.
 decodeIOPortionWith :: Peek a -> BS.ByteString -> IO (Offset, a)
 decodeIOPortionWith mypeek (BS.PS x s len) =
     withForeignPtr x $ \ptr0 ->
         let ptr = ptr0 `plusPtr` s
-            end = ptr `plusPtr` len
-         in do
-             (ptr2, x') <- runPeek mypeek end ptr
-             if ptr2 > end
-                 then throwIO $ PeekException (end `minusPtr` ptr2) "Overshot end of buffer"
-                 else return (ptr2 `minusPtr` ptr, x')
+        in decodeIOPortionWithFromPtr mypeek ptr len
 {-# INLINE decodeIOPortionWith #-}
 
+-- | Like 'decodeIOWith', but using 'Ptr' and length instead of a 'ByteString'.
+decodeIOWithFromPtr :: Peek a -> Ptr Word8 -> Int -> IO a
+decodeIOWithFromPtr mypeek ptr len = do
+    (offset, x) <- decodeIOPortionWithFromPtr mypeek ptr len
+    let remaining = len - offset
+    if remaining > 0
```

Comment:
I agree a simple comparison of the length and offset would be simpler. I think the reason it's written the way it is is that the `decodePortion` functions already check for consuming too much input (which you never want to allow), and the `decode` functions just add a check for consuming less output than possible.

Just for clarification: do you suggest to get rid of the `decodePortion` type of functions and do just one check, or just to replace the `remaining > 0` check with a `len /= offset` check?